src/ssh/allowed_signer.rs

```diff
@@ -0,0 +1,218 @@
+use std::collections::VecDeque;
+use std::fmt;
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+
+use super::pubkey::PublicKey;
+use crate::{error::Error, Result};
+
+/// A type which represents an allowed signer entry.
+/// Please refer to [ssh-keygen-1.ALLOWED_SIGNERS] for more details about the format.
+/// [ssh-keygen-1.ALLOWED_SIGNERS]: https://man.openbsd.org/ssh-keygen.1#ALLOWED_SIGNERS
+#[derive(Debug, PartialEq, Eq)]
+pub struct AllowedSigner {
+    /// A list of principals, each in the format USER@DOMAIN.
+    pub principals: Vec<String>,
+
+    /// Indicates that this key is accepted as a CA.
+    /// This is a standard option.
+    pub cert_authority: bool,
+
+    /// Specifies a list of namespaces that are accepted for this key.
+    /// This is a standard option.
+    ///
+    /// Note: The specification allows spaces inside double quotes. However, this is not supported.
+    pub namespaces: Option<Vec<String>>,
+
+    /// Time at or after which the key is valid.
+    /// This is a standard option.
+    pub valid_after: Option<u64>,
+
+    /// Time at or before which the key is valid.
+    /// This is a standard option.
+    pub valid_before: Option<u64>,
+
+    /// Public key of the entry.
+    pub key: PublicKey,
+}
+
+/// A type which represents a collection of allowed signer entries.
+/// Please refer to [ssh-keygen-1.ALLOWED_SIGNERS] for more details about the format.
+/// [ssh-keygen-1.ALLOWED_SIGNERS]: https://man.openbsd.org/ssh-keygen.1#ALLOWED_SIGNERS
+#[derive(Debug, PartialEq, Eq)]
+pub struct AllowedSigners {
+    /// A collection of allowed signers
+    pub allowed_signers: Vec<AllowedSigner>,
+}
+
+impl AllowedSigner {
+    /// Parse an allowed signer entry from a given string.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use sshcerts::ssh::AllowedSigner;
+    ///
+    /// let allowed_signer = AllowedSigner::from_string(concat!(
+    ///     "user@domain.tld ecdsa-sha2-nistp384 AAAAE2VjZHNhLXNoYTItbmlzdHAzODQAAAAIbmlzdHAzODQAAABhBGJe",
+    ///     "+IDGRhlQdDp+/AIsTXGaVhWaQUbHJwqLDlQIh7V4xatO6E/4Uva+f70WzxgM7xHPGUqafqNAcxxVBP4jkx3HVDRSr7C3",
+    ///     "NVBpr0ZaKXu/hFiCo/4kry4H5MGMEvKATA=="
+    /// )).unwrap();
+    /// println!("{:?}", allowed_signer);
+    /// ```
+    pub fn from_string(s: &str) -> Result<AllowedSigner> {
+        let mut parts: VecDeque<&str> = s.split_whitespace().collect();
+
+        // An allowed signer must contian at least 3 parts: principals, key type and pubkey data
+        if parts.len() < 3 {
+            return Err(Error::InvalidFormat);
+        }
+
+        let principals = parts.pop_front().ok_or(Error::InvalidFormat)?;
+        let principals: Vec<&str> = principals.split(',').collect();
+        let principals = principals.iter().map(|s| s.to_string()).collect();
```

Comment:
Do we enforce a format for principals?

Reply:
Seems like this is not done in other APIs (`cert.rs`) so maybe we don't need this?

Reply:
I think principals are pretty open. There might be characters which cannot be in principals (","?)

Reply:
I'm not enforcing format on principals